### PR TITLE
Federation support

### PR DESCRIPTION
**Pull request by fancycode**
Needs https://github.com/nextcloud/spreed/pull/12604

### CODE REVIEW

File: federation.go [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+						if j == remoteSessionId {
+							msg.Event.Leave[idx] = localSessionId
+							if c.closeOnLeave.Load() {
+								doClose = true
+							}
+							break
+						}
+					}
+				}
+			}
```
Comment (danxuliu) [inline federation.go:491]: Events with `target: room` can have `type: message` and include a roomid (see, for example, the received signaling message when posting a chat message), but that roomid is currently not converted.

Comment (fancycode) [reply federation.go]: Thanks, changed.

**Comment by fancycode** [comment]
@danxuliu this is ready for merging from my side, could you please do a test with your Talk changes and see if everything is working as expected? Thanks!

**Comment by SystemKeeper** [comment]
I played around with it today, what I noticed: 
* When I first join the signaling controller without federated information
* Then join the signaling controller with federated information (so still the same conversation)

The second join kills the first session on the Nextcloud server, so joining a call then fails, because I don't have an active session on the nextcloud server.
While I can see that it _might_ be intended this way, it was totally unexpected to me. Is it intended or should the second join just "overwrite" the first one (without killing the session in Nextcloud?)?

The second thing I noticed, when I join the same room twice fast (this time both requests do have federated information attached), I sometimes got "no_such_room" back and only the second join worked.

**Comment by SystemKeeper** [comment]
> * When I first join the signaling controller without federated information
> * Then join the signaling controller with federated information (so still the same conversation)
> 
> The second join kills the first session on the Nextcloud server, so joining a call then fails, because I don't have an active session on the nextcloud server.

It seems unrelated to the fact with/without federated information, when I try to join the same server again, a leave is sent to the nextcloud instance and then my join request fails and I get brute-force protected on top. While I can work around some cases, we still can end up in the case where a duplicated join happens (that's why we added `already_joined` back then). Is there a chance to make that more resilient?

**Comment by fancycode** [comment]
@SystemKeeper duplicate federated join requests are now handled in https://github.com/strukturag/nextcloud-spreed-signaling/pull/776/commits/5d4efe8cd6d8994a98e58d143f25fa5a6f857a87

**Comment by SystemKeeper** [comment]
> @SystemKeeper duplicate federated join requests are now handled in [5d4efe8](https://github.com/strukturag/nextcloud-spreed-signaling/commit/5d4efe8cd6d8994a98e58d143f25fa5a6f857a87)

Tested and works, thanks!

**Comment by SystemKeeper** [comment]
Is it expected, that the capabilities expire so often?

```
2024-07-25 19:28:13 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:13 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:13 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEGSND9JoFDbD2tyMoAfkJ3jQazY9p
2024-07-25 19:28:13 wHZN7o7X1rVB4GfKuWT2qHbCGDU69MRDm/oV92MKJcFCbSCQwPecm/uUAw==
2024-07-25 19:28:13 -----END PUBLIC KEY-----
2024-07-25 19:28:13  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:13 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:13 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:13 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:13 -----END PUBLIC KEY-----
2024-07-25 19:28:13  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:15 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:15 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:15 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:15 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:15 -----END PUBLIC KEY-----
2024-07-25 19:28:15  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:15 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:15 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:15 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEGSND9JoFDbD2tyMoAfkJ3jQazY9p
2024-07-25 19:28:15 wHZN7o7X1rVB4GfKuWT2qHbCGDU69MRDm/oV92MKJcFCbSCQwPecm/uUAw==
2024-07-25 19:28:15 -----END PUBLIC KEY-----
2024-07-25 19:28:15  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:22 client.go:353: Client 4XW9g09SsUOgfo-vWddARFlRx0RiTch2JKHAtBlpAdp8blN0bjJ1VGl6bER4NHFQem8ta3Z4cFZiWXFmQS1OVk55THJKMHQ0T1RWRlpYZTVYTndpUXFnM2RicGY2bkphaDFBTTBZbjJob3RuY085Vk1KbUxUeXJxbGpkVGEyOVdaSmx2UDhsakVXVWtrOU16WmdUeHhyT0xvdjJmOGRMb1RveWFqOXFXVUxxR21wRUF5eVhvdzBmR0pJS1JFeC1QazkweDVDUlllfDQ5MzgyOTEyNzE= has RTT of 0 ms (738.334µs)
2024-07-25 19:28:22 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:22 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:23 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:23 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:23 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:23 -----END PUBLIC KEY-----
2024-07-25 19:28:23  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:23 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
...
```

**Comment by fancycode** [comment]
The expiration honors any caching headers set (since #780) instead of using a hardcoded cache duration of one hour. The default is one minute if no caching headers are returned.

Could you please check the response when you query the capabilities manually?
```
curl -ik -H "OCS-ApiRequest: true" -H "Accept: application/json" https://server.domain.tld/ocs/v2.php/cloud/capabilities
```

Anyway, if something is not working correctly, please open a new issue as this is unrelated to the federation changes.

**Comment by SystemKeeper** [comment]
> please open a new issue as this is unrelated to the federation changes.

Will do, sorry, looked related to this PR, as it didn't happen before, but I missed #780.

**Pull request merged**